From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: KatanaMajesty <sasha.panasenk@gmail.com>
Date: Fri, 1 Jul 2022 17:02:17 +0300
Subject: [PATCH] Added Bloom Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 56e261efa654e4a6872ccea28f0461df13845d13..21921e63c4e80dce855063b5647f40a155cddafe 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1937,6 +1937,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Bloom start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getBloomConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Bloom end
+
         /**
          * Sends the component to the player
          *
